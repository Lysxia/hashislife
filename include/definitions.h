#ifndef DEFINITIONS_H
#define DEFINITIONS_H

#include <stdint.h>

/*! \defgroup rules Game rules */
/*!@{*/
//! Rule set encoded as a bit pattern
/*! 18 bits are used to indicate Born/Stay transitions */
/*! \todo Explain */
typedef uint32_t rule; // at least 18 bits

#define CONWAY ((rule) 6152) // "b3/s23";
/*!@}*/

char digit_to_char(int d);
int itoa(char *dest, int src, int base);

#define ALIVE_CELL_CHAR 'O'
#define  DEAD_CELL_CHAR '.'

#endif
