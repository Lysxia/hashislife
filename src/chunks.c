#include <stdlib.h>
#include "chunks.h"

//! Create single block
ChunksBlocks *chunksblocks_new(size_t n)
{
  ChunksBlocks *cb = malloc(sizeof(ChunksBlocks));

  if ( cb == NULL )
    return NULL;

  cb->chunksb_size = 0;
  cb->chunksb_block = malloc(n);
  cb->chunksb_next = NULL;

  if ( cb->chunksb_block == NULL ) {
    free(cb);
    return NULL;
  }
  else
    return cb;
}

/*! In case of failure, the `.chunks_blocks` field is set to `NULL`
*/
Chunks chunks_new(const size_t data_size, const int max_len)
{
  Chunks c = {
    .chunks_data_size = data_size,
    .chunks_max_len = max_len,
    .chunks_blocks = chunksblocks_new(data_size * max_len) };

  return c;
}

/*! Do not call twice on a successfully created chunk,
  or ever on a failed one.
*/
void chunks_delete(Chunks c)
{
  ChunksBlocks *cb = c.chunks_blocks;
  while ( cb != NULL )
  {
    ChunksBlocks *cb_next = cb->chunksb_next;
    free(cb->chunksb_block);
    free(cb);
    cb = cb_next;
  }
}

/*! Allocates a unit of memory */
void *chunks_alloc(Chunks c)
{
  if (    (unsigned) c.chunks_blocks->chunksb_size
       == c.chunks_data_size * c.chunks_max_len ) // Current block is full
  {
    /* Create new block
      A little hack to get around the fact that c is passed by value:
      the new block takes the fields of the filled block,
      while the latter is cleared for a new one...
      I wonder how much one gains over a pointer argument.
    */
    ChunksBlocks *cb = malloc(sizeof(ChunksBlocks));

    if ( cb == NULL )
      return NULL;

    char *block = malloc(c.chunks_data_size * c.chunks_max_len);

    if ( block == NULL ) {
      free(cb);
      return NULL;
    }

    *cb = *c.chunks_blocks;

    c.chunks_blocks->chunksb_size = 0;
    c.chunks_blocks->chunksb_block = block;
    c.chunks_blocks->chunksb_next = cb;
  }

  void *p
    = &c.chunks_blocks->chunksb_block[c.chunks_blocks->chunksb_size];
  c.chunks_blocks->chunksb_size += c.chunks_data_size;
  return p;
}

