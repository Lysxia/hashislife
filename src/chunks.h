#ifndef CHUNKS_H
#define CHUNKS_H
/*! \defgroup chunks Memory chunks */
/*!@{*/
typedef struct Chunks Chunks;
typedef struct ChunksBlocks ChunksBlocks;

//! Managed blocks
struct Chunks {
  const size_t  chunks_data_size;
  const int     chunks_max_len;
  ChunksBlocks *chunks_blocks;
};

//! Single block
struct ChunksBlocks {
  int           chunksb_size;
  char         *chunksb_block;
  ChunksBlocks *chunksb_next;
};

Chunks chunks_new(const size_t data_size, const int max_len);
void chunks_delete(Chunks);
void *chunks_alloc(Chunks);
/*!@}*/
#endif
