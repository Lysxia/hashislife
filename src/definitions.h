#ifndef DEFINITIONS_H
#define DEFINITIONS_H

#include <stdint.h>
typedef uint32_t rule; // at least 18 bits

int itoa(char *dest, int src);

#endif
