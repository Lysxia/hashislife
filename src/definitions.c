#include <assert.h>
#include "definitions.h"

char digit_to_char(int d) {
  if ( d < 10 )
    return '0' + d;
  else if ( d < 36 )
    return 'A' + d - 10;
  else
    return '?';
}

int itoa(char *dest, int src, int base) {
  assert( base < 36 );
  int i = 0, j;
  do
  {
    dest[i++] = digit_to_char(src % base);
    src /= base;
  }
  while ( src > 0 );

  for ( j = 0 ; j < i / 2 ; j++ )
  {
    char tmp = dest[i-j-1];
    dest[i-j-1] = dest[j];
    dest[j] = tmp;
  }

  dest[i] = '\0';

  return i;
}
