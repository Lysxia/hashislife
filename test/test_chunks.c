#include <stdio.h>
#include <stdlib.h>
#include "definitions.h"
#include "chunks.h"

void show_chunk(Chunks c) {
  printf("sizeof: %zu, length: %d\n", c.chunks_data_size, c.chunks_max_len);
  ChunksBlocks *cb;
  for ( cb = c.chunks_blocks ; cb != NULL ; cb = cb->chunksb_next ) {
    int i;
    for ( i = 0 ; i < cb->chunksb_size ; i++ ) {
      if ( i % c.chunks_data_size == 0 && i != 0 )
        printf(" ");
      printf("%c%c", digit_to_char(cb->chunksb_block[i]/16),
                     digit_to_char(cb->chunksb_block[i]%16));
    }
    printf("\n");
  }
}

int main(void) {
  const int n = 5;
  Chunks c = chunks_new(sizeof(int), n);
  show_chunk(c);
  int i, k, j = 0;
  for ( k = 0 ; k < 7 ; k++ ) {
    for ( i = 0 ; i < 3 ; i++, j++ ) {
      * (int *) chunks_alloc(c) = j;
    }
    show_chunk(c);
  }
  return 0;
}

