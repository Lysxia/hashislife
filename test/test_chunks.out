sizeof: 4, length: 5

sizeof: 4, length: 5
00000000 01000000 02000000
sizeof: 4, length: 5
05000000
00000000 01000000 02000000 03000000 04000000
sizeof: 4, length: 5
05000000 06000000 07000000 08000000
00000000 01000000 02000000 03000000 04000000
sizeof: 4, length: 5
0A000000 0B000000
05000000 06000000 07000000 08000000 09000000
00000000 01000000 02000000 03000000 04000000
sizeof: 4, length: 5
0A000000 0B000000 0C000000 0D000000 0E000000
05000000 06000000 07000000 08000000 09000000
00000000 01000000 02000000 03000000 04000000
sizeof: 4, length: 5
0F000000 10000000 11000000
0A000000 0B000000 0C000000 0D000000 0E000000
05000000 06000000 07000000 08000000 09000000
00000000 01000000 02000000 03000000 04000000
sizeof: 4, length: 5
14000000
0F000000 10000000 11000000 12000000 13000000
0A000000 0B000000 0C000000 0D000000 0E000000
05000000 06000000 07000000 08000000 09000000
00000000 01000000 02000000 03000000 04000000
